### PR TITLE
Stop simplifying ancestor merges

### PR DESCRIPTION
**Pull request by martinvonz**
# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

**Comment by martinvonz** [comment]
@ilyagr FYI

File: lib/src/rewrite.rs [hunk 1]
```diff
@@ -241,9 +241,21 @@ pub enum EmptyBehaviour {
 // change the RebaseOptions construction in the CLI, and changing the
 // rebase_commit function to actually use the flag, and ensure we don't need to
 // plumb it in.
-#[derive(Clone, Default, PartialEq, Eq, Debug)]
+#[derive(Clone, PartialEq, Eq, Debug)]
 pub struct RebaseOptions {
     pub empty: EmptyBehaviour,
+    /// If a merge commit would end up with one parent being an ancestor of the
+    /// other, then filter out the ancestor.
+    pub no_ancestor_merge: bool,
```
Comment (yuja) [inline lib/src/rewrite.rs:249]: nit: I prefer `simplify_ancestor_merge` or something like that. `no_` sounds like a flag to disable extra operation.

File: lib/src/rewrite.rs [hunk 2]
```diff
@@ -553,16 +565,23 @@ impl<'settings, 'repo> DescendantRebaser<'settings, 'repo> {
...
+        let new_parent_ids: Vec<_> = if self.options.no_ancestor_merge {
+            let head_set: HashSet<_> = self
+                .mut_repo
+                .index()
+                .heads(&mut new_parent_ids.iter())
+                .into_iter()
+                .collect();
+            new_parent_ids
+                .into_iter()
+                .filter(|new_parent| head_set.contains(new_parent))
```
Comment (yuja) [inline lib/src/rewrite.rs:569]: nit: (at your option) `.retain()` could be used, but that would make the vec mutable.

**Review by yuja — APPROVED** [review]
lgtm, but ilyagr might know some pitfalls.

File: lib/src/rewrite.rs (hunk 1 above)
Comment (martinvonz) [reply lib/src/rewrite.rs]: Done.

File: lib/src/rewrite.rs (hunk 2 above)
Comment (martinvonz) [reply lib/src/rewrite.rs]: Done.

**Comment by martinvonz** [comment]
@ilyagr, I'll merge this so we get ~2 weeks to test it before the next release.

**Comment by ilyagr** [comment]
Sorry I was slow. I still hope to look it over.

**Comment by martinvonz** [comment]
> Sorry I was slow. I still hope to look it over.

No problem. It seemed pretty straight-forward in theory, so I'm just hoping to test it more in practice.

**Pull request merged**